rust/user-info-fetcher/src/main.rs

```diff
@@ -0,0 +1,262 @@
+        path: PathBuf,
+    },
+
+    #[snafu(display("unable to parse config file"))]
+    ParseConfig { source: serde_json::Error },
+
+    #[snafu(display("failed to parse listen address"))]
+    ParseListenAddr { source: AddrParseError },
+
+    #[snafu(display("failed to register SIGTERM handler"))]
+    RegisterSigterm { source: std::io::Error },
+
+    #[snafu(display("failed to run server"))]
+    RunServer { source: hyper::Error },
+
+    #[snafu(display("failed to construct http client"))]
+    ConstructHttpClient { source: reqwest::Error },
+
+    #[snafu(display("failed to read ca certificate"))]
+    ReadCaCert { source: std::io::Error },
+
+    #[snafu(display("failed to parse ca certificate"))]
+    ParseCaCert { source: reqwest::Error },
+}
+
+async fn read_config_file(path: &Path) -> Result<String, StartupError> {
+    tokio::fs::read_to_string(path)
+        .await
+        .context(ReadConfigFileSnafu { path })
+}
+
+#[tokio::main]
+async fn main() -> Result<(), StartupError> {
+    let args = Args::parse();
+
+    stackable_operator::logging::initialize_logging(
+        "OPA_OPERATOR_LOG",
+        APP_NAME,
+        args.common.tracing_target,
+    );
+
+    let shutdown_requested = tokio::signal::ctrl_c().map(|_| ());
+    #[cfg(unix)]
+    let shutdown_requested = {
+        let mut sigterm = tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate())
+            .context(RegisterSigtermSnafu)?;
+        async move {
+            let sigterm = sigterm.recv().map(|_| ());
+            pin_mut!(shutdown_requested, sigterm);
+            future::select(shutdown_requested, sigterm).await;
+        }
+    };
+
+    let config = Arc::<crd::Config>::new(
+        serde_json::from_str(&read_config_file(&args.config).await?).context(ParseConfigSnafu)?,
+    );
+    let credentials = Arc::new(match &config.backend {
+        // TODO: factor this out into each backend (e.g. when we add LDAP support)
+        crd::Backend::None {} => Credentials {
+            username: "".to_string(),
+            password: "".to_string(),
+        },
+        crd::Backend::Keycloak(_) => Credentials {
+            username: read_config_file(&args.credentials_dir.join("clientId")).await?,
+            password: read_config_file(&args.credentials_dir.join("clientSecret")).await?,
+        },
+    });
+
+    let mut client_builder = ClientBuilder::new();
+
+    // TODO: I'm not so sure we should be doing all this keycloak specific stuff here.
+    // We could factor it out in the provider specific implementation (e.g. when we add LDAP support).
+    // I know it is for setting up the client, but an idea: make a trait for implementing backends
+    // The trait can do all this for a genric client using an implementation on the trait (eg: get_http_client() which will call self.uses_tls())
+    if let crd::Backend::Keycloak(keycloak) = &config.backend {
+        if keycloak.tls.uses_tls() && !keycloak.tls.uses_tls_verification() {
```

Comment:
This is more of a complaint about the op-rs API, but this feels really misuse-prone to me. If we want clients to branch from this then it would be preferable to have something like:

```
match tls.verification() {
    None => { ... }
    Some(TlsVerification::Custom { ca_mount_path }) => { ... }
    Some(TlsVerification::System) => { ... }
}
```

Reply:
To be clear, fixing that is way out of scope for this PR

Reply:
Agree. Will close this here, and it will come up when it needs to.